### PR TITLE
Suggest adding an array length if possible

### LINKED ISSUE #100448: "using `_` for array lengths is unstable" error should suggest the number of elements
Given the following code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=fe35b2f5f585f0303661ac61e880d1fe

```rust
const NUMBERS: [i32; _] = [
    1, 2, 3
];
```

The current output is:

```
error[[E0658]](https://doc.rust-lang.org/stable/error-index.html#E0658): using `_` for array lengths is unstable
 --> src/lib.rs:1:22
  |
1 | const NUMBERS: [i32; _] = [
  |                      ^
  |
  = note: [see issue #85077 <https://github.com/rust-lang/rust/issues/85077>](https://github.com/rust-lang/rust/issues/85077) for more information

error: in expressions, `_` can only be used on the left-hand side of an assignment
 --> src/lib.rs:1:22
  |
1 | const NUMBERS: [i32; _] = [
  |                      ^ `_` not allowed here

For more information about this error, try `rustc --explain E0658`.
error: could not compile `playground` due to 2 previous errors
```

Ideally, the output should suggest that this should be [i32; 3].

### PR DESCRIPTION
**Pull request by TaKO8Ki**
fixes #100448

### CODE REVIEW

File: compiler/rustc_typeck/src/check/expr.rs [hunk 1]
```diff
@@ -1307,7 +1307,39 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                _ => None,
+            };
+            if let Some(ty) = ty
+                && let hir::TyKind::Array(_, length) = ty.kind
+                && let hir::ArrayLen::Body(hir::AnonConst { hir_id, .. }) = length
+                && let Some(span) = self.tcx.hir().opt_span(hir_id)
+            {
+                match self.tcx.sess.diagnostic().steal_diagnostic(span, StashKey::UnderscoreForArrayLengths) {
+                    Some(mut err) => {
+                        err.span_suggestion_verbose(
```
Comment (compiler-errors) [inline compiler/rustc_typeck/src/check/expr.rs:1331]: nit: does this need to be verbose? I think `help: consider specifying the array length: 3` is fine inline.

File: compiler/rustc_typeck/src/check/expr.rs [hunk 2]
```diff
@@ -1307,7 +1307,39 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                            span,
+                            "consider adding an array length",
```
Comment (compiler-errors) [inline compiler/rustc_typeck/src/check/expr.rs:1333]: nit: "consider specifying the array length" might be better, though I don't really care either way

**Review by compiler-errors — COMMENTED** [review]
cool!

**Comment by compiler-errors** [comment]
anyways r=me after nits are addressed or not

@bors delegate+

File: src/test/ui/array-slice-vec/suggest-array-length.stderr [hunk 3]
```diff
@@ -0,0 +1,48 @@
+error: in expressions, `_` can only be used on the left-hand side of an assignment
```
Comment (WaffleLapkin) [inline src/test/ui/array-slice-vec/suggest-array-length.stderr:1]: Why are all errors in this file duplicated? For example this is and the one at line 28 has the same cause.

**Review by WaffleLapkin — COMMENTED** [review]
Annoyingly this also doesn't work with references to arrays, for example:
```rust
const REPORT_DESCR: &[u8; _] = &[1]
```
won't trigger the suggestion.

**Pull request merged**